### PR TITLE
Use intersection types in Object.assign defintion

### PR DESCRIPTION
**Pull request by mhegazy**

### CODE REVIEW

File: src/lib/es6.d.ts [hunk 1]
```diff
@@ -125,6 +125,33 @@ interface ObjectConstructor {
...
+      */
+    assign<T, U, V>(target: T, source1: U, source2: V): T & U & V;
+
+    /**
+      * Copy the values of all of the enumerable own properties from one or more source objects to a 
+      * target object. Returns the target object.
+      * @param target The target object to copy to.
+      * @param source1 first source objects to copy properties from.
+      * @param source2 second source objects to copy properties from.
+      * @param source3 second source objects to copy properties from.
```
Comment (DanielRosenwasser) [inline src/lib/es6.d.ts]: third

Comment (DanielRosenwasser) [inline src/lib/es6.d.ts]: "source objects" -> "source object"

"to copy properties from" -> "from which to copy properties"

File: src/lib/es6.d.ts [hunk 2]
```diff
@@ -125,6 +125,33 @@ interface ObjectConstructor {
...
+      * @param source2 second source objects to copy properties from.
+      */
+    assign<T, U, V>(target: T, source1: U, source2: V): T & U & V;
+
+    /**
+      * Copy the values of all of the enumerable own properties from one or more source objects to a 
+      * target object. Returns the target object.
+      * @param target The target object to copy to.
+      * @param source1 first source objects to copy properties from.
+      * @param source2 second source objects to copy properties from.
```
Comment (DanielRosenwasser) [inline src/lib/es6.d.ts]: "source objects" -> "source object"

"to copy properties from" -> "from which to copy properties"

File: src/lib/es6.d.ts [hunk 3]
```diff
@@ -125,6 +125,33 @@ interface ObjectConstructor {
...
+      */
+    assign<T, U, V>(target: T, source1: U, source2: V): T & U & V;
+
+    /**
+      * Copy the values of all of the enumerable own properties from one or more source objects to a 
+      * target object. Returns the target object.
+      * @param target The target object to copy to.
+      * @param source1 first source objects to copy properties from.
```
Comment (DanielRosenwasser) [inline src/lib/es6.d.ts]: "source objects" -> "source object"

"to copy properties from" -> "from which to copy properties"

File: src/lib/es6.d.ts [hunk 4]
```diff
@@ -125,6 +125,33 @@ interface ObjectConstructor {
...
+      * @param source2 second source objects to copy properties from.
+      * @param source3 second source objects to copy properties from.
+      */
+    assign<T, U, V, W>(target: T, source1: U, source2: V, source3: W): T & U & V & W;
+
+    /**
+      * Copy the values of all of the enumerable own properties from one or more source objects to a
```
Comment (DanielRosenwasser) [inline src/lib/es6.d.ts]: "own properties" -> "properties on the object itself"

Comment (mhegazy) [inline src/lib/es6.d.ts]: i think own properties has reference to JS hasOwnProeprties and getOwnProeprties.. etc.. so i would say it is more meaningful in this sense.

File: src/lib/es6.d.ts (hunk 1 above)
Comment (JsonFreeman) [inline src/lib/es6.d.ts]: Er... "to copy properties from" is indeed better I believe.

File: src/lib/es6.d.ts (hunk 2 above)
Comment (JsonFreeman) [inline src/lib/es6.d.ts]: "to copy properties from"

File: src/lib/es6.d.ts (hunk 3 above)
Comment (JsonFreeman) [inline src/lib/es6.d.ts]: "to copy properties from" is the better syntax.

File: src/lib/es6.d.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/lib/es6.d.ts]: Yeah, agreed, sorry

File: src/lib/es6.d.ts [hunk 5]
```diff
@@ -125,7 +125,34 @@ interface ObjectConstructor {
...
+      */
+    assign<T, U, V, W>(target: T, source1: U, source2: V, source3: W): T & U & V & W;
+
+    /**
+      * Copy the values of all of the enumerable own properties from one or more source objects to a 
+      * target object. Returns the target object.
+      * @param target The target object to copy to.
+      * @param sources One or more source objects from which to copy properties
       */
     assign(target: any, ...sources: any[]): any;
```
Comment (DanielRosenwasser) [inline src/lib/es6.d.ts]: If you make this take a `T` and `U` array, then users can specifically pass in a union type for their `U`

Comment (ahejlsberg) [inline src/lib/es6.d.ts]: @DanielRosenwasser But then the resulting type would follow the pattern T & (U1 | U2 | U3 ... ) which I don't think is what you want.

Comment (JsonFreeman) [inline src/lib/es6.d.ts]: You'd need something like the proposal in #3870.

File: src/lib/es6.d.ts (hunk 4 above)
Comment (vladima) [inline src/lib/es6.d.ts]: Current message is aligned with [Object.assign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) from MDN so I'll keep it as is

**Comment by vladima** [comment]
:+1:

**Comment by mhegazy** [comment]
merged manually.

**Pull request closed without merging**